### ISSUE TITLE
QKNorm ablation

### ISSUE THREAD

**Issue by Niccolo-Ajroldi** [OpenEuroLLM/Taskboard#72]
We want to assess the benefits of QKNorm in combination with a higher LR on the 1.7B model.

We can start by following the setup from https://github.com/OpenEuroLLM/Taskboard/issues/16.

**Comment by Niccolo-Ajroldi** [comment]
@spyysalo we were. thinking of reproducing https://github.com/OpenEuroLLM/Taskboard/issues/16 in its latest stage, but with Nemotron-cc, exploring different combinations of LR's.

**Comment by aaronkl** [comment]
After discussing with @Niccolo-Ajroldi and @geoalgo, we came up with the following experiment setup:

We will explore a successive halving approach across 3 different model scales. We start with a large grid on the smallest scale and successively prune the configuration to run on larger scales:
- tiny model: 300BT , 130M params, seq_len 2048
- small model: 300BT, 400M params, seq_len 2048
- larger model: 300BT, 1.7B params, seq_len 2048

We can cool down multiple intermediate checkpoints to estimate the correlation of hyperparameters across token budgets.


Search space: 

- bsz: {2M, 4M} 
- qknorm : {true, false}
- lr-warmup: {100BT / 4 tokens, 100BT / 2 tokens, 100BT tokens} (30% of 300BT (used in open-sci reference runs = 100 BT)
- LR: {MIN, MAX}

**Comment by Niccolo-Ajroldi** [comment]
Quick update on this task with some results!

## Goal
We want to investigate if *long warmup*, *high learing rate*, and *qknorm* are beneficial.

## Setting
We train a 130M model, on 300BT Nemotron-HQ-synth-mix data. We follow the OpenSci setup, with a couple of exceptions: we use a seq_len of 2k instead of 4k, and a global_bsz of 1000 instead of 1008.
We also follow Sampo's [LUMI setup](https://github.com/spyysalo/lumi-fineweb-replication/blob/main/train-gpt.sh) for the following hyperparameters:`hidden_dropout=0.0`, `attention_dropout=0.0`, `attention_softmax_in_fp32=True`, `add_bias_linear=False`.

## 1st Sweep
We explore the following hyparameter combinations:
- QKNORM : {true, false}
- LR_WARMUP_PCT: 3%, 10%, 30% (of tot tokens)
- LR: 3e-4, 3e-3
We are also repeating the experiment with double the bsz (4M instead of 2M).

## Results
Here is a plot with LR=3e-4 (we plot a single LR for better visualization, nevertheless the darkest blue line is overlapped with a red one and is barely visible). We report rolling mean and std of LM loss.
<img width="7069" height="1988" alt="Image" src="https://github.com/user-attachments/assets/7fe39903-6857-469b-a6fa-e4cb6f8336c5" />

Loss has very high oscillations and differences are subtle, but we can still draw some conclusions.

We draw all the lines in the sweep with a slightly higher averaging window and report the mean only.
<img width="7091" height="1988" alt="Image" src="https://github.com/user-attachments/assets/3f5a6835-0778-451c-b153-90bbe1b00387" />

## Comment
- **large vs small LR**: a 10x larger learning rate (solid line) performs better than a smaller LR (dotted line) at the beginning, then it underperforms in the stable regime of the WSD schedule, finally giving better loss after cooling down. I personally find this behavior to be consistent across many LM experiments I ran for other projects on different architectures and datasets.
- **QKNORM**: enabling QKNORM (red lines) yields better results wrt to not enabling it (blue lines). This is consistent across LR's and warmup durations. The performance gain does not seem to increase when using a 10x larger LR. 
- **WARMUP**: the lighter the color, the longer the warmup. We find that a longer warmup does not produce any benefit, and slightly decreases performance. We don't find any evidence that a longer warmup might be beneficial when using a larger LR.

**Comment by JeniaJitsev** [comment]
In open-sci-ref, we did hyperparam tuning on larger model scale, 1.3B. There, longer warmup was clearly beneficial for both 50B and 300B token budgets. It might be as the training instability increases with model scale, slower warmup becomes more and more important, while on small scales like 0.13B, it has no effect.

**Comment by Niccolo-Ajroldi** [comment]
Posting results for larger batch size (bsz=2000, whereas we previously used 1000)

When increasing the batch size, we keep the number of tokens fixed, which effectively halves the number of steps.
We keep the sequence length fixed to 2048.
We also specify the warmup duration as a fraction of the total tokens, so for a larger batch size this means a warmup phase with less steps compared with the largr bsz.

## QKNorm, Warmup, LR
<img width="7090" height="1988" alt="Image" src="https://github.com/user-attachments/assets/86585b73-5095-4261-9879-ccadc29d43e6" />

Similar results as before:
- qknorm is aslways beneficial
- higher LR yields better result
- 30% warmup is useless
At LR=3e-3 we observe some instabilities at the beginning of training, both with and without QKNorm.


## Comparing runs with different batch sizes
We can also compare the validation loss profiles with shorter and larger bsz.

We can visualize Loss vs Steps:
<img width="7077" height="1988" alt="Image" src="https://github.com/user-attachments/assets/d151a350-e3ef-438b-8092-23e562e77239" />

Or Loss vs Tokens:
<img width="7077" height="1988" alt="Image" src="https://github.com/user-attachments/assets/cdb1a019-c5a0-4bab-8c8e-51cbcd335199" />

We observe a slight advantage in training at a smaller batch size of 1000 samples wrt 2000. This suggests the critical batch size is smaller than 2000. The gains are however small, and the largest bsz allows for higher parallelization and hence shorter training time.
